ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,604 @@
+    let highest_block = observer
+        .provider()
+        // we get finalized block to avoid block reorganization
+        // because we restore the database only for the latest block of a chain,
+        // and thus the reorganization can lead us to an empty block
+        .get_block(BlockId::Number(BlockNumberOrTag::Finalized))
+        .await
+        .context("failed to get latest block")?
+        .expect("latest block always exist");
+    let highest_block = H256(highest_block.header.hash.0);
+
+    log::info!("Syncing chain head {highest_block}");
+    observer.force_sync_block(highest_block).await?;
+    while let Some(event) = observer.next().await {
+        match event? {
+            ObserverEvent::Blob(_blob) => {
+                unreachable!("no blob events should occur before chain head is synced")
+            }
+            ObserverEvent::Block(_) => {}
+            ObserverEvent::BlockSynced(data) => {
+                debug_assert_eq!(highest_block, data.block_hash);
+                break;
+            }
+        }
+    }
+
+    Ok(highest_block)
+}
+
+async fn sync_from_network(
+    network: &mut NetworkService,
+    db: &Database,
+    program_states: &BTreeMap<ActorId, H256>,
+    latest_block_header: &BlockHeader,
+) -> Schedule {
+    let mut schedule_restorer = ScheduleRestorer::new(latest_block_header.height);
+
+    let program_states: HashMap<H256, HashSet<ActorId>> =
+        program_states
+            .iter()
+            .fold(HashMap::new(), |mut acc, (&program_id, &state)| {
+                acc.entry(state).or_default().insert(program_id);
+                acc
+            });
+    let mut manager = RequestManager::default();
+    for (&state, program_ids) in &program_states {
+        manager.add(state, RequestMetadata::ProgramState { program_ids });
+    }
+
+    loop {
+        let wait_for_network = manager.request(network, db);
+
+        let (completed, pending) = manager.stats();
+        log::info!("[{completed:>05} / {pending:>05}] Getting network data");
+
+        if wait_for_network {
+            let stream = network.filter_map(|event| async move {
+                if let NetworkEvent::DbResponse { request_id, result } = event {
+                    Some((request_id, result))
+                } else {
+                    None
+                }
+            });
+            let mut stream = pin!(stream);
+            let (request_id, result) = stream
+                .next()
+                .await
+                .expect("network service stream is infinite");
+
+            match result {
+                Ok(response) => {
+                    manager.handle_response(request_id, response, db);
+                }
+                Err((request, err)) => {
+                    network.db_sync().retry(request);
+                    log::warn!("{request_id:?} failed: {err}. Retrying...");
+                }
+            }
+        }
```

Comment:
Why this logic is not inside manager.take_responses() ? It should be asynchronous then and taking reference to network.